code_generation/producer.py

```diff
@@ -45,6 +46,13 @@ def __init__(self, name, call, input, output, scopes):
         log.debug("| scopes: {}".format(self.scopes))
         log.debug("-----------------------------------------")
 
+    def check_output(self):
+        if self.output != None and not self.output_checked:
+            for scope in self.scopes:
```

Comment:
do not loop over scopes here. Take the scope by argument and only apply for that scope. The producer might be used multiple times in different pipelines such that this function will be called a second time and crash. Also, if the producer is not run in all scopes it is designed for, you will erroneously mark the quantity as available in that scope.